### PR TITLE
Get utils ready for better Agent torch support.

### PR DESCRIPTION
**Pull request by sven1977**
## Why are these changes needed?

Add torch-specific functions to utils

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by sven1977** [comment]
cc'ing @ericl for review.

File: rllib/agents/trainer_template.py [hunk 1]
```diff
@@ -171,7 +167,7 @@ def with_updates(**overrides):
         """
         return build_trainer(**dict(original_kwargs, **overrides))
 
-    trainer_cls.with_updates = with_updates
+    trainer_cls.with_updates = staticmethod(with_updates)
```
Comment (ericl) [inline rllib/agents/trainer_template.py]: Any reason not to decorate the original method?

File: rllib/utils/__init__.py [hunk 2]
```diff
@@ -42,63 +28,31 @@ class new_base(mixins.pop(), base):
...
-        return None
-
-    try:
-        import tensorflow_probability as tfp
-        return tfp
-    except ImportError:
-        return None
-
-
 __all__ = [
```
Comment (ericl) [inline rllib/utils/__init__.py]: Thanks for cleaning this up

File: rllib/utils/explained_variance.py [hunk 3]
```diff
@@ -2,12 +2,18 @@
...
 
 tf = try_import_tf()
+torch, nn = try_import_torch()
 
 
-def explained_variance(y, pred):
-    _, y_var = tf.nn.moments(y, axes=[0])
-    _, diff_var = tf.nn.moments(y - pred, axes=[0])
-    return tf.maximum(-1.0, 1 - (diff_var / y_var))
+def explained_variance(y, pred, backend="tf"):
```
Comment (ericl) [inline rllib/utils/explained_variance.py]: We've been using the term 'framework' instead of backend, would be good to be consistent with the terminology.

File: rllib/utils/backend.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
+from __future__ import absolute_import
```
Comment (ericl) [inline rllib/utils/backend.py]: Consider 'framework.py' for consistency.

File: rllib/utils/numpy.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
+from __future__ import absolute_import
+from __future__ import division
```
Comment (ericl) [inline rllib/utils/numpy.py]: What's this for?

**Review by ericl — CHANGES_REQUESTED** [review]

File: rllib/agents/trainer_template.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/trainer_template.py]: PyCharm marks it as bad style due to the decorator being inside a function, not a class.

File: rllib/utils/numpy.py (hunk 5 above)
Comment (sven1977) [reply rllib/utils/numpy.py]: Test cases: Numpy-based checks vs simple tf/torch model forward passes for e.g. loss function tests.

File: rllib/utils/backend.py (hunk 4 above)
Comment (sven1977) [reply rllib/utils/backend.py]: Ok, makes sense. Will rename.

File: rllib/utils/numpy.py (hunk 5 above)
Comment (sven1977) [reply rllib/utils/numpy.py]: e.g. you create a Model with a very simple NN (1 fc layer w/ 10 linear nodes). Then simulate the forward pass, check the Model's output against its weights using this module's `fc()`-function and then calculate the exact expected loss value and  validate it with the loss functions actual output.

Comment (sven1977) [reply rllib/utils/numpy.py]: Note: The next PR will have test cases for PG- and PPO losses (both torch and tf).

Comment (ericl) [reply rllib/utils/numpy.py]: Hm looks fine to merge, though isn't it a bit like testing TF itself?

File: rllib/utils/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+from __future__ import print_function
+
+import numpy as np
+
+from ray.rllib.utils.framework import try_import_tf
+
+tf = try_import_tf()
+
+
+def check(x, y, decimals=5, atol=None, rtol=None, false=False):
```
Comment (ericl) [inline rllib/utils/test_utils.py]: Consider `check_equals`

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Looks good pending changes.

File: rllib/utils/test_utils.py (hunk 6 above)
Comment (sven1977) [reply rllib/utils/test_utils.py]: Ok, will rename this in the next PR.

**Comment by ericl** [comment]
I think test_actor is flaking on the osx build, so this is just pending jenkins finishing.
